File: src/Core/Mixin.fs
namespace Zeta.Core

open System
open System.Runtime.CompilerServices

/// Thread-safe and GC-safe weak-keyed identity table for attaching state to objects.
type WeakMap<'Key, 'Value when 'Key : not struct and 'Key : not null and 'Value : not struct>() =
    let table = ConditionalWeakTable<'Key, 'Value>()

    /// Attach a state value to the key. Overwrites if it already exists.
    member _.Set(key: 'Key, value: 'Value) =
        table.AddOrUpdate(key, value)

    /// Try to get the state value associated with the key.
    member _.TryGet(key: 'Key) : 'Value option =
        match table.TryGetValue(key) with
        | true, value -> Some value
        | false, _ -> None

    /// Get the state value associated with the key, or construct it using the factory if absent.
    member _.Get(key: 'Key, factory: 'Key -> 'Value) : 'Value =
        table.GetValue(key, ConditionalWeakTable<'Key, 'Value>.CreateValueCallback(factory))

    /// Delete the entry associated with the key. Returns true if removed, false otherwise.
    member _.Delete(key: 'Key) : bool =
        table.Remove(key)

File: src/Core.Python/src/zeta/mixin.py
import weakref
from typing import TypeVar

TKey = TypeVar("TKey")
TValue = TypeVar("TValue")


class WeakMap[TKey, TValue]:
    """
    GC-safe weak-keyed identity table for attaching state to objects.
    """

    def __init__(self) -> None:
        self._dict: weakref.WeakKeyDictionary[TKey, TValue] = (
            weakref.WeakKeyDictionary()
        )

    def set(self, key: TKey, value: TValue) -> None:
        """
        Attach a state value to the key. Overwrites if it already exists.
        """
        self._dict[key] = value

    def get(self, key: TKey) -> TValue | None:
        """
        Try to get the state value associated with the key. Returns None if absent.
        """
        return self._dict.get(key)

    def delete(self, key: TKey) -> bool:
        """
        Delete the entry associated with the key. Returns True if removed, False otherwise.
        """
        if key in self._dict:
            del self._dict[key]
            return True
        return False
